src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java

```diff
@@ -76,93 +92,166 @@ private void deleteWebhooks(List<BitbucketWebhook> webhooks) {
                         .collect(toList());
+
+        //Creating lists of different kinds of webhooks
         List<BitbucketWebhook> webhookWithMirrorSync = ownedHooks.stream()
-                .filter(hook -> hook.getEvents().contains(MIRROR_SYNCHRONIZED_EVENT.getEventId()))
+                .filter(hook -> hook.getEvents().contains(MIRROR_SYNCHRONIZED.getEventId()))
                 .collect(toList());
-        List<BitbucketWebhook> webhookWithRepoRefChange = ownedHooks
+        List<BitbucketWebhook> webhookWithRepoRefChangeOnly = ownedHooks
                 .stream()
-                .filter(hook -> hook.getEvents().contains(REPO_REF_CHANGE.getEventId()))
+                .filter(hook -> hook.getEvents().equals(Collections.singleton(REPO_REF_CHANGE.getEventId())))
                 .collect(toList());
 
-        if (ownedHooks.size() == 0 ||
-            (webhookWithMirrorSync.size() == 0 && event == MIRROR_SYNCHRONIZED_EVENT) ||
-            (webhookWithRepoRefChange.size() == 0 && event == REPO_REF_CHANGE)) {
-            BitbucketWebhookRequest webhook = createRequest(request, event);
+        List<BitbucketWebhook> webhookWithPROnly = ownedHooks
+                .stream()
+                .filter(hook -> hook.getEvents().equals(ALL_PULL_REQUEST_EVENT_IDS))
+                .collect(toList());
+
+        List<BitbucketWebhook> webhookWithRepoRefChangeOrPR = ownedHooks
+                .stream()
+                .filter(hook -> hook.getEvents().stream().anyMatch(refAndPREventIds::contains))
+                .collect(toList());
+
+        //Determining whether to create a new webhook
+        //We will handle mirror syncs separate from all other events.
+        // If no webhook with wanted events exist, we register a new one.
+        if (ownedHooks.isEmpty() ||
+            (webhookWithMirrorSync.isEmpty() && events.contains(MIRROR_SYNCHRONIZED)) ||
+            (webhookWithRepoRefChangeOrPR.isEmpty() && (events.contains(PULL_REQUEST_OPENED) || events.contains(REPO_REF_CHANGE)))) {
+            BitbucketWebhookRequest webhook = createRequest(request, events);
             BitbucketWebhook result = webhookClient.registerWebhook(webhook);
             LOGGER.info("New Webhook registered - " + result);
             return result;
         }
 
         BitbucketWebhook mirrorSyncResult =
-                handleExistingWebhook(request, webhookWithMirrorSync, MIRROR_SYNCHRONIZED_EVENT);
+                handleExistingWebhook(request, webhookWithMirrorSync, Collections.singleton(MIRROR_SYNCHRONIZED));
 
-        BitbucketWebhook repoRefResult = handleExistingWebhook(request, webhookWithRepoRefChange, REPO_REF_CHANGE);
+        //Determining which existing webhooks to update/delete for ref changes and pull request events
+        BitbucketWebhook repoResult;
+        Collection<BitbucketWebhookEvent> supportedEvents = new HashSet<>();
+        if (events.contains(REPO_REF_CHANGE) && events.contains(PULL_REQUEST_OPENED)) {
```

Comment:
Why does it only check against `PULL_REQUEST_OPENED`? I may be misunderstanding, but shouldn't it check whether `events` contains any pull request event?

Reply:
This is checking if we should perform a build, which we do when a push is made to a repo or a PR is opened. I notice that this conditional _doesn't_ check for source branch updated events (presumably because they're covered by repo ref change), but I think this would be more correct:
 ```suggestion
        if (events.stream().anyMatch(event -> refAndPREventIds.contains(event.getEventId()))) {
```